### ISSUE TITLE
CLI to publish/subscribe with `iox2` from the terminal

### ISSUE THREAD

**Issue by orecham** [eclipse-iceoryx/iceoryx2#413]
## Brief feature description

Make it possible to publish, subscribe to iceoryx2 services directly from the CLI, e.g.

```console
iox2 service publish <TOPIC> <PAYLOAD>
iox2 service subscribe <TOPIC>
```

## Detailed information

Nice to have:
1. Make it possible to plug in custom logic for working with payloads, e.g.
    1. Publish data from a file
    2. Display image payloads from subscribers directly in the terminal

## Related

#790 - notifier/listener with `iox2`

**Comment by brosier01** [comment]
Hi @orecham, I'm currently working on this feature. First, I'm trying to implement the publish subcommand. I'm using the “publish_subscribe_subscriber” example as a test endpoint. As I understand it, in order to communicate with the subscriber, I need to define a service with a specific payload size. Can I get the size of this payload and create a new service with this payload size at runtime? (Because at compile time, I don't know the payload size and the subscriber needs a payload of size)

**Comment by orecham** [comment]
@brosier01 

Hmm, this is a tough problem to solve - something I completely overlooked when imagining this feature, my mistake.

As a first step, I guess the CLI could be made to only use [the `[u8]` payload type](https://github.com/eclipse-iceoryx/iceoryx2/blob/main/examples/rust/publish_subscribe_dynamic_data/publisher.rs#L23) so that concrete types aren't needed. We are currently working on updating `iceoryx2` so that the [`maximum_slice_len`](https://github.com/eclipse-iceoryx/iceoryx2/blob/main/examples/rust/publish_subscribe_dynamic_data/publisher.rs#L29) does not need to be provided when defining services of type `[u8]`. This however would mean that applications would also need to use the `[u8]` payload type for the CLI to connect to it, which is not ideal for many reasons (usability, safety, etc.). Proceeding with this first though would allow for focusing on the other (maybe simpler) aspects of implementing the CLI.

To get the CLI to work with services with [concrete types](https://github.com/eclipse-iceoryx/iceoryx2/blob/main/examples/rust/publish_subscribe/publisher.rs#L25), two ideas come to mind (but I think they might definitely need some more thought):

1. Add functionality for `iceoryx2` applications to compile type definitions into a separate shared library
    1. The CLI would need to have a way to find these libraries to get the type definitions
        1. Ideally this would be automatic, but could be done manually by passing an argument as a stepping stone
1. Use something like [`schemars`](https://crates.io/crates/schemars) to store type descriptions along with services?
    1. Not sure how feasible it is, but could also be an interesting option
    2. Not sure if this removes the requirement of using `[u8]` payload types though

Let me know if you have any other ideas. I am not yet super confident that one of these two options would be optimal. Might need some discussion before we work out what makes the most sense.

> Can I get the size of this payload and create a new service with this payload size at runtime?

Until we remove the `max_slice_len` requirement, you could maybe just make the size the actual size of the payload.

**Comment by elfenpiff** [comment]
@brosier01 First of all, thanks for taking care of the CLI it is much appreciated!

And here some constructive help for your challenge.

First of all, you have to acquire the type details for the payload and the user header that are stored in the static service details of the service. To do this, you first need to acquire the static service details with:

```rust
let details = ipc::Service::details(
    &"My/Funk/ServiceName".try_into()?,
    Config::global_config(),
    MessagingPattern::PublishSubscribe,
).unwrap().unwrap();
```

When you have it you need to extract the static details for publish subscribe
```rust
let user_header;
let payload;
if let iceoryx2::service::static_config::messaging_pattern::MessagingPattern::PublishSubscribe(
    config,
) = details.static_details.messaging_pattern()
{
    user_header = config.message_type_details().user_header.clone();
    payload = config.message_type_details().payload.clone();

    println!("user header: {:?}", user_header);
    println!("payload: {:?}", payload);
} else {
    panic!()
}
```

With those details you can now define a service with custom type details. This is not safe since you could mix things up here and induce undefined behavior or a crash in another application. It is also internal API, but we use the same approach for our internal tooling so I think the CLI can be an exception too. To be able to override the payload and user header type details you need to define the `[CustomPayloadMarker]` as payload type and `CustomHeaderMarker` as user header type.

```rust
use iceoryx2::service::builder::publish_subscribe::{CustomHeaderMarker, CustomPayloadMarker};

let expert_service = unsafe {
    node.service_builder(&"My/Funk/ServiceName".try_into()?)
        .publish_subscribe::<[CustomPayloadMarker]>()
        .user_header::<CustomHeaderMarker>()
        .__internal_set_payload_type_details(&payload)
        .__internal_set_user_header_type_details(&user_header)
        .open()
        .unwrap();
};
```

The subscriber and the publisher port now behave like if they would have been created with `[u8]` payload type and you are able to send and receive any arbitrary data.

**Comment by brosier01** [comment]
Hi @elfenpiff, I've tried to implement your code in my CLI tool but my current application crashes. Here is the code : 

```rust
let node = NodeBuilder::new().create::<ipc::Service>()?;

let expert_service = unsafe {
        node.service_builder(&"My/Funk/ServiceName".try_into()?)
        .publish_subscribe::<[CustomPayloadMarker]>()
            .user_header::<CustomHeaderMarker>()
            .__internal_set_payload_type_details(&payload)
            .__internal_set_user_header_type_details(&user_header)
            .open()
            .unwrap();
    };
```

As I understand it, it's the service_builder method that's causing the error.

**Comment by elfenpiff** [comment]
@brosier01 can you please set the log level to tracing `set_log_level(LogLevel::Trace)` and add the full output that leads to the crash?

**Comment by elfenpiff** [comment]
@brosier01 In the output line 18 shows the problem:

```
Unable to open publish subscribe service since the service offers the type "MessageTypeDetails { header: TypeDetail { variant: FixedSize, type_name: "iceoryx2::service::header::publish_subscribe::Header", size: 32, alignment: 8 } ... which is not compatible to the requested type "MessageTypeDetails { header: TypeDetail { variant: FixedSize, type_name: "iceoryx2::service::header::publish_subscribe::Header", size: 24, alignment: 8 } ...".
```

For whatever reason the `Header`, not the user header, has for one variant a size of 24 and for the other a size of 32. The header which is used for publish subscribe is located in `iceoryx2/src/service/header/publish_subscribe.rs` and has the size 24 with alignment 8.

But I have no idea why this does not work on your side - do you have maybe anywhere a typo? I will look in parallel if the implementation is maybe somewhere screwed up.

**Comment by elfenpiff** [comment]
@brosier01 We have an internal introspection client and there we use this code snippet successfully:

```rust
if let iceoryx2::service::static_config::messaging_pattern::MessagingPattern::PublishSubscribe(pub_sub_details) = pub_sub_details.static_details.messaging_pattern() {
   let message_type = pub_sub_details.message_type_details();

   subscriber = unsafe { node
       .service_builder(name)
       .publish_subscribe::<[CustomPayloadMarker]>()
       .user_header::<CustomHeaderMarker>()
       .__internal_set_payload_type_details(&message_type.payload)
       .__internal_set_user_header_type_details(&message_type.user_header)
       .open()
       .unwrap();
}
```

**Comment by brosier01** [comment]
@elfenpiff applogies, I mixed up the versions of iceoryx2. Now it works fine on my side. From this expert service can I create a new pubisher like the publisher.rs in the examples?

**Comment by elfenpiff** [comment]
@brosier01 Yes you can create both ports when you have opened/created a service like I described above. 
FYI: in the previous example the variable name is wrong, it should be `service` instead of `subscriber`

```rust
   service = unsafe { node // service not subscriber
       .service_builder(name)
       .publish_subscribe::<[CustomPayloadMarker]>()
       .user_header::<CustomHeaderMarker>()
       .__internal_set_payload_type_details(&message_type.payload)
       .__internal_set_user_header_type_details(&message_type.user_header)
       .open()
       .unwrap();
```

**Comment by brosier01** [comment]
Hi @elfenpiff, I'm trying to send data from a publisher : 
```rust
let service = unsafe { node
            .service_builder(&"My/Funk/ServiceName".try_into()?)
            .publish_subscribe::<[CustomPayloadMarker]>()
            .user_header::<CustomHeaderMarker>()
            .__internal_set_payload_type_details(&payload)
            .__internal_set_user_header_type_details(&user_header)
            .open().unwrap()
        };

let publisher = service.publisher_builder().create().unwrap();

let sample = unsafe {publisher.loan_custom_payload(1)?};
```

How can I write my data into my payload and send it ?

Sorry I'm a beginner with Iceoryx2

**Comment by elfenpiff** [comment]
@brosier01 

> Sorry I'm a beginner with Iceoryx2

Don't feel sorry. We are here to help you - this is the beauty of an open-source community :) 

> How can I write my data into my payload and send it ?

When you call `loan_custom_payload(N)` you get `N` times the size of the `payload`. So assume your payload has a size of 4 and `N` is 2 you get 8 bytes since it is handled like a slice.

Let's assume your data type is `Fuu` you could write the data like:
```rust
unsafe { (sample.payload_mut()[0].as_mut_ptr() as *mut Fuu).write(Fuu::new()) };
```

If you want to write arbitrary bytes into it that were provided as arguments from the console you can do it like:
```rust
for n in 0..payload.size { // unsure if it is really payload.size, but you require the type size of the payload
  unsafe { (sample.payload_mut()[n].as_mut_ptr() as *mut u8).write(123) };
}
```

**Comment by elfenpiff** [comment]
@brosier01 

Those failures are not a problem. `[T]` = trace log output, `[D]` debug, `[I]` info, `[W]` warning, `[E]` error, `[F]` fatal last log output before a crash.

In this case it just means that a specific shared memory could not be opened but iceoryx2 could handle it by creating it and then open it.
`[D]` is always emitted when a function returns a `Result` with an error.

For the user only are only warnings and errors interesting. Just add
```rust
set_log_level(LogLevel::Warn);
```
at the top of your program and the log messages vanishes.

**Comment by elfenpiff** [comment]
@brosier01 I think the problem is that your subscriber is waiting for messages and the publisher is being created, sends out the message and then destroys itself - making it undetectable for the subscriber since it is happening so fast.
So could you please do the following:
1. start the subscriber and call in an infinite loop `receive`
2. create the publisher, send the sample and then wait a second

Do you now receive the sample on subscriber side? If so, I need to add a little something extra to the API so that publisher blocks until the subscriber has confirmed the connection.

**Comment by brosier01** [comment]
@elfenpiff you're right! After a one-second wait, the subscriber received the data.
Another point for this CLI tool: did I need to know the payload data structure ? IIn my case, I'm sending the TransmissionData data structure, which is composed of {i32, i32, f64}. With the CLI tool iox-service details, I can obtain information on the payload : 

payload: (
                variant: FixedSize,
                type_name: "examples_common::transmission_data::TransmissionData",
                size: 16,
                alignment: 8,
)

But I still don't know how to fill my payload.

**Comment by elfenpiff** [comment]
@brosier01 This is no longer your problem, but the user has to solve it. You already have ensured that the size and alignment fits and now the user has to fill the sample. My idea was that the user either provides a binary stream so that the cli could be handled somehow like this - via a unix pipe

```sh
cat /dev/some_device | iox2 service publish some_topic
```

Or that the user can provide the bytes explicitly as hex codes like
```sh
iox2 service publish --hex 0xab 0x12 0x32 0x44 ...
```

For the unix pipe approach, you fill the sample until it is full and send it out, for the hex code approach you have to check if the correct amount of bytes was provided and then send it out.

**Comment by brosier01** [comment]
@elfenpiff does the hex option mean it has several payload formats ? Xhat is your vision? What formats do I need to implement ?

**Comment by elfenpiff** [comment]
@brosier01 no, it is still strongly typed. But the hex approach is the only possible way to send any data format via command line, except the piping approach.

So the underlying publisher would just translate the hex code into byte code, copies it directly into the sample and on the other side it can be interpreted as an int or struct. The cli user is then responsible for providing a correct hex format.

**Comment by brosier01** [comment]
If hex is the only option. Is the cli "--hex" option mandatory ?

**Comment by orecham** [comment]
@brosier01 @elfenpiff 

For completion, the idea I was considering was supporting functionality similar to what is available in ROS 2 (as in example [found here](https://roboticsbackend.com/ros2-topic-cmd-line-tool-debug-ros2-topics-from-the-terminal/#ros2_topic_pub_%E2%80%93_Publish_to_a_topic_from_the_terminal)):

```
$ ros2 topic pub -r 10 /greetings example_interfaces/msg/String "{data: 'Hello from terminal'}"
publisher: beginning loop
publishing #1: example_interfaces.msg.String(data='Hello from terminal')

publishing #2: example_interfaces.msg.String(data='Hello from terminal')

...
```

This is the direction I was thinking in with my [previous comment](https://github.com/eclipse-iceoryx/iceoryx2/issues/413#issuecomment-2493559770). The benefit of this is that it could open up avenues for users to quickly interact with a running system from the CLI. The scenario I have in mind is a developer plugging their laptop into the running system for debugging / testing.

This functionality is more complex to implement than sending the (unchecked) hex bytes though, since the CLI would need a way to find the message definition. For the first PR, feel free to choose just one of the possibilities (whatever is easiest or most interesting).
